src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -1581,19 +1581,23 @@ private void ReportResultsToSubmission(BuildResult result)
                 {
                     BuildSubmission submission = _buildSubmissions[result.SubmissionId];
 
-                    submission.CompleteResults(result);
-
-                    // If the request failed because we caught an exception from the loggers, we can assume we will receive no more logging messages for
-                    // this submission, therefore set the logging as complete. IntrnalLoggerExceptions are unhandled exceptions from the logger. If the logger author does
-                    // not handle an exception the eventsource wraps all exceptions (except a logging exception) into an internal logging exception.
-                    // These exceptions will have their stack logged on the commandline as an unexpected failure. If a logger author wants the logger
-                    // to fail gracefully then can catch an exception and log a LoggerException. This has the same effect of stopping the build but it logs only
-                    // the exception error message rather than the whole stack trace.
-                    if (result.Exception is InternalLoggerException || result.Exception is LoggerException || result.Exception is InvalidOperationException)
+                    /* If the request failed because we caught an exception from the loggers, we can assume we will receive no more logging messages for
+                     * this submission, therefore set the logging as complete. InternalLoggerExceptions are unhandled exceptions from the logger. If the logger author does
+                     * not handle an exception the eventsource wraps all exceptions (except a logging exception) into an internal logging exception.
+                     * These exceptions will have their stack logged on the commandline as an unexpected failure. If a logger author wants the logger
+                     * to fail gracefully then can catch an exception and log a LoggerException. This has the same effect of stopping the build but it logs only
+                     * the exception error message rather than the whole stack trace.
+                     * 
+                     * If any other exception happened and logging is not completed, then go ahead and completed now since this is the last place to do it.
+                     * Otherwise the submission would remain uncompleted and the build would hang.
+                    */
+                    if (!submission.LoggingCompleted && result.Exception != null)
```

Comment:
What's the stack for the cases where this was misbehaving? Your explanation sounds entirely reasonable but I'd like to check it out a bit further.

Reply:
This can manifest badly in multiple ways, since the submission is never marked as completed and lingers around in the BuildManager state forever:
- API users get blocked on `BuildSubmission.WaitHandle`
- API users never get pinged on their `BuildSubmissionCompleteCallback`
- BuildManager.EndBuild blocks forever waiting for the submission to complete